src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest`1.cs

```diff
@@ -318,6 +327,178 @@ private void VerifyDiagnosticResults(IEnumerable<Diagnostic> actualResults, Immu
             }
         }
 
+        private ImmutableArray<(Diagnostic? actual, DiagnosticResult? expected)> MatchDiagnostics(Diagnostic[] actualResults, DiagnosticResult[] expectedResults)
+        {
+            expectedResults = expectedResults.ToOrderedArray();
+
+            var bestMatchCount = actualResults.Length + expectedResults.Length;
+            var bestMatch = actualResults.Select(result => ((Diagnostic?)result, default(DiagnosticResult?))).Concat(expectedResults.Select(result => (default(Diagnostic?), (DiagnosticResult?)result))).ToImmutableArray();
+
+            var builder = ImmutableArray.CreateBuilder<(Diagnostic? actual, DiagnosticResult? expected)>();
+            bool[] usedExpected = new bool[expectedResults.Length];
+            RecursiveMatch(0, 0, 0, usedExpected);
+
+            return bestMatch;
+
+            // Returns the minimum number of unmatched items
+            int RecursiveMatch(int firstActualIndex, int firstExpectedIndex, int unmatchedActualResults, bool[] usedExpected)
```

Comment:
This method might not be changed much but it's quite difficult to understand it all (or I am too stupid xD) it might be worth having a few more comments on the various key parts/points of the algorithm.

Reply:
It's a recursive backtracking algorithm, implemented in a relatively naive way. I'll add some comments.

Reply:
Comments are added